### PR TITLE
Add optional TLS spec for RouteGroups

### PR DESCRIPTION
**Pull request by rickhlx**
This will allow Skipper to add TLS certs defined in RouteGroups to the certificate registry which will be used to terminate TLS based on the configured hostnames. This behaviors mimics [Ingress Spec](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/ingress-v1/#IngressSpec).

### CODE REVIEW

**Comment by szuecs** [comment]
@rickhlx will check this as soon https://github.com/zalando/skipper/pull/2814 is merged.

File: zalando.org_routegroups.yaml
Comment (AlexanderYastrebov) [inline zalando.org_routegroups.yaml:228]: Where does this pattern come from?

File: apis/zalando.org/v1/types.go
Comment (AlexanderYastrebov) [inline apis/zalando.org/v1/types.go]: I understand this is copied from kubectl explain ingress but "name/s" is strange and its unclear what "tlsSecret" references to. Would it bepossible to rephrase somehow?

File: zalando.org_routegroups.yaml
Comment (szuecs) [reply zalando.org_routegroups.yaml:228]: Yeah would be good to know. It looks also a bit wrong.
No *.example possible but you can have example

Comment (rickhlx) [reply zalando.org_routegroups.yaml:228]: This was generated automatically from `make zalando.org_routegroups.yaml`

Comment (AlexanderYastrebov) [reply zalando.org_routegroups.yaml:228]: Ok, I was wondering where it comes from because I did not see any pattern defined in types.go
I turns our we have a perl snippet that applies to this field as well by luck:
https://github.com/szuecs/routegroup-client/blob/98ae2ff4e8c830d90f13f2a9fd626034eb057a39/Makefile#L33-L37

Comment (AlexanderYastrebov) [reply zalando.org_routegroups.yaml:228]: @szuecs I think its will handle domains properly, see https://regex101.com/r/DjRVGZ/1

Comment (AlexanderYastrebov) [reply zalando.org_routegroups.yaml:228]: > we have a perl snippet that applies to this field

Its a hack anyways and it works so I guess its fine.

File: apis/zalando.org/v1/types.go
Comment (AlexanderYastrebov) [reply apis/zalando.org/v1/types.go]: I mean there is no `tlsSecret` anywhere but there is `secretName`.
I also think `name(s)` reads better than `name/s`.
`Host specifies` -> `Hosts specify`

File: apis/zalando.org/v1/types.go
Comment (AlexanderYastrebov) [inline apis/zalando.org/v1/types.go]: Its not clear from the user perspective what happens when `secretName` is omitted. The implementation https://github.com/zalando/skipper/pull/2814 ignores hosts when secret is not specified. Maybe we should make `secretName` required since "TLS routing based on SNI hostname alone" will work anyways even when whole "tls" is omitted (assuming "-kubernetes-enable-tls" is set and some other ingress/routegroup registers certs for the same spec.hosts)

File: apis/zalando.org/v1/types.go
Comment (AlexanderYastrebov) [inline apis/zalando.org/v1/types.go]: The port does not have to be 443 so I think we should omit "on port 443".

File: zalando.org_routegroups.yaml
Comment (AlexanderYastrebov) [reply zalando.org_routegroups.yaml:228]: I looked again at https://github.com/kubernetes-sigs/controller-tools/issues/342 which is still not fixed. One workaround mentioned is to introduce type alias but that would be a breaking change unfortunately: 

```diff
diff --git a/Makefile b/Makefile
index 619a1a3..5db97a0 100644
--- a/Makefile
+++ b/Makefile
@@ -32,9 +32,6 @@ $(GENERATED): $(CRD_TYPE_SOURCE)
 
 $(GENERATED_CRD): go.mod $(GENERATED)
        go run sigs.k8s.io/controller-tools/cmd/controller-gen crd:crdVersions=v1 paths=./apis/... output:crd:dir=.
-       # workaround to add pattern to array items. Not supported by controller-gen
-       # ref: https://github.com/kubernetes-sigs/controller-tools/issues/342
-       perl -i -p0e 's|(\s*)(hosts:.*?items:)|$$1$$2$$1    pattern: "^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*\$$"|sg' $(GENERATED_CRD)
 
 build.local: $(LOCAL_BINARIES) $(GENERATED_CRD)
 build.linux: $(LINUX_BINARIES) $(GENERATED_CRD)
diff --git a/apis/zalando.org/v1/types.go b/apis/zalando.org/v1/types.go
index a45de6a..a1e0883 100644
--- a/apis/zalando.org/v1/types.go
+++ b/apis/zalando.org/v1/types.go
@@ -39,10 +39,13 @@ type RouteGroupList struct {
        Items []RouteGroup `json:"items"`
 }
 
+// +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+type Hostname string
+
 // +k8s:deepcopy-gen=true
 type RouteGroupSpec struct {
        // List of hostnames for the RouteGroup
-       Hosts []string `json:"hosts,omitempty"`
+       Hosts []Hostname `json:"hosts,omitempty"`
        // List of backends that can be referenced in the routes
        Backends []RouteGroupBackend `json:"backends"`
        // DefaultBackends is a list of default backends defined if no explicit
diff --git a/apis/zalando.org/v1/zz_generated.deepcopy.go b/apis/zalando.org/v1/zz_generated.deepcopy.go
index eea8969..d9ece11 100644
--- a/apis/zalando.org/v1/zz_generated.deepcopy.go
+++ b/apis/zalando.org/v1/zz_generated.deepcopy.go
@@ -187,7 +187,7 @@ func (in *RouteGroupSpec) DeepCopyInto(out *RouteGroupSpec) {
        *out = *in
        if in.Hosts != nil {
                in, out := &in.Hosts, &out.Hosts
-               *out = make([]string, len(*in))
+               *out = make([]Hostname, len(*in))
                copy(*out, *in)
        }
        if in.Backends != nil {
diff --git a/cli/rg-client-test/main.go b/cli/rg-client-test/main.go
index 60802d4..4c3394d 100644
--- a/cli/rg-client-test/main.go
+++ b/cli/rg-client-test/main.go
@@ -41,7 +41,7 @@ func main() {
        // test create
        name := "myrg"
        namespace := "default"
-       hostname := "myrg.teapot-e2e.zalan.do"
+       hostname := rgv1.Hostname("myrg.teapot-e2e.zalan.do")
        port := 83
        newRg := &rgv1.RouteGroup{
                ObjectMeta: metav1.ObjectMeta{
@@ -49,7 +49,7 @@ func main() {
                        Namespace: namespace,
                },
                Spec: rgv1.RouteGroupSpec{
-                       Hosts: []string{hostname},
+                       Hosts: []rgv1.Hostname{hostname},
                        Backends: []rgv1.RouteGroupBackend{
                                {
                                        Name:        name,
diff --git a/zalando.org_routegroups.yaml b/zalando.org_routegroups.yaml
index bd78b61..55cd9b3 100644
--- a/zalando.org_routegroups.yaml
+++ b/zalando.org_routegroups.yaml
@@ -143,7 +143,7 @@ spec:
               hosts:
                 description: List of hostnames for the RouteGroup
                 items:
-                  pattern: "^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+                  pattern: ^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
                   type: string
                 type: array
               routes:
```

File: apis/zalando.org/v1/types.go
Comment (AlexanderYastrebov) [reply apis/zalando.org/v1/types.go]: So should we make `hosts` and `secretName` both required?

Comment (rickhlx) [reply apis/zalando.org/v1/types.go]: Added the `kubebuilder` "Required" validation and re-generated the CRD.

File: apis/zalando.org/v1/types.go
Comment (AlexanderYastrebov) [inline apis/zalando.org/v1/types.go]: I think you just need to drop `omitempty` like we do for other string fields instead of using annotation.

Comment (rickhlx) [reply apis/zalando.org/v1/types.go]: Looks better now, thanks for that.

**Comment by AlexanderYastrebov** [comment]
LGTM, will merge after we land https://github.com/zalando/skipper/pull/2814 (please update CRD there)

File: zalando.org_routegroups.yaml
Comment (szuecs) [reply zalando.org_routegroups.yaml:228]: > @szuecs I think its will handle domains properly, see https://regex101.com/r/DjRVGZ/1

It can't do `*.example.org`, which is ok for certificates.

I think you miss the difference between domain names and cert SAN entries. You can also have an IP in SAN and Hosts is something else than tls hosts.

Comment (AlexanderYastrebov) [reply zalando.org_routegroups.yaml:228]: The implementation will check that value is also in spec.hosts which has this pattern.
We can rethink or relax pattern later but for now I think its fine. Or what do you suggest?

Comment (szuecs) [reply zalando.org_routegroups.yaml:228]: ok

**Review by szuecs — APPROVED** [review]
thanks for the PR

**Pull request merged**
